waiter/src/waiter/service_description.clj

```diff
@@ -1252,22 +1253,24 @@
   [reference-type->stale-fn references]
   (every? (fn [reference] (stale-reference? reference-type->stale-fn reference)) references))
 
-(defn service-id->idle-timeout
-  "Computes the idle timeout, in minutes, for a given service.
-   If the service is active or was created by on-the-fly, the idle timeout is retrieved from the service description.
-   Else, the idle timeout is the sum of the fallback period seconds and the stale service timeout."
+(defn service->gc-time
+  "Computes the time when a given service should be GC-ed.
+   If the service has GC disabled (by configuring idle-timeout-mins=0), nil is returned.
+   If the service is active or was created by on-the-fly, the GC time is retrieved from the service description.
+   Else, the GC time is the sum of the fallback period seconds and the stale service timeout."
   [service-id->service-description-fn service-id->references-fn token->token-parameters reference-type->stale-fn
-   attach-token-defaults-fn service-id]
+   attach-token-defaults-fn service-id last-modified-time]
   (let [{:strs [idle-timeout-mins]} (service-id->service-description-fn service-id)
         references (service-id->references-fn service-id)]
     (if (service-references-stale? reference-type->stale-fn references)
       (let [{:strs [stale-timeout-mins]} (attach-token-defaults-fn {})]
-        (log/info service-id "that uses tokens is stale")
+        (log/info service-id "that uses references is stale")
         (if-let [source-tokens (->> references (map :token) (remove nil?) (map :sources) seq)]
-          (source-tokens->idle-timeout token->token-parameters attach-token-defaults-fn source-tokens)
+          (source-tokens->gc-time token->token-parameters attach-token-defaults-fn source-tokens)
           ;; use the default token stale timeout for a stale service built without tokens
-          stale-timeout-mins))
-      idle-timeout-mins)))
+          (t/plus last-modified-time (t/minutes stale-timeout-mins))))
+      (when (pos? idle-timeout-mins)
```

Comment:
Let's use `cond` here instead of `if`+`when`.

`(if a b (when c d))` → `(cond a b c d)`

Note that `cond` returns `nil` when none of the conditions match:
https://clojuredocs.org/clojure.core/cond